### ISSUE TITLE
Timeout issues when signing certs with step-ca and PKSC#11+yubihsm2

### ISSUE THREAD

**Issue by mikansontap123** [smallstep/certificates#937]
### Subject of the issue
I have built step-ca with CGO to enable PKCS#11 support. My instance is configured to use a yubihsm2 module. 

Generating keys via step-pkcs11-init works fine. The issue comes when submitting a request for a certificate and getting it signed. The first attempt always works - a signed certificate is delivered back to me. A second attempt, if done within <15 seconds of the first attempt, also works as expected. However, a second attempt that is done more than >15 seconds after the first attempt always fails, and any subsequent attempts will also fail until step-ca is restarted. 

### Your environment
* OS - Debian 11/bullseye
* Version - step-ca 0.19.0 (built from CGO)

### Steps to reproduce
`step ssh certificate certSubject id_ecdsa --ca-url https://my-step-ca.local --root root_ca.crt`
Run the above. Then run the same command again more than 15 seconds after the first attempt. 

### Expected behaviour
The above command should always return a signed certificate (assuming valid parameters are given to the provisioner.)

### Actual behaviour
After the first attempt, which is successful, if any further attempts are made after >15 seconds, the following error is returned:
`ERRO[0068] duration=7.387314ms duration-ns=7387314 error="authority.SignSSH: error signing certificate: error signing certificate: pkcs11: 0x6: CKR_FUNCTION_FAILED"`

If I restart step-ca, the next signing attempt will again work without an issue, and any further attempts fail if 15 seconds has elapsed. 

### Additional context
15 seconds is the default keepalive value when a session is established with yubihsm2. While the error output makes no direct reference to any keepalive or timeouts, I have a suspicion there is a relation here as it consistently fails after >15 seconds has elapsed, and consistently works if a signing attempt is made within <15 seconds. 

I am still able to access the HSM manually with yubihsm-shell after the error occurs, so it does not appear that the yubihsm-connector service is failing - it is just the interaction between step-ca and the yubihsm that no longer works, as simply restarting step-ca allows a new certificate to get signed.

**Comment by maraino** [comment]
Hi @mikansontap123, which version of the yubihsm2-sdk are you using? I've just tested with the one I have installed, 2021.08, it is a little behind, but I haven't seen any problems.

Can you also enable in `YUBIHSM_PKCS11_CONF` `debug` and `libdebug`, and optionally `dinout`. Something like:

```
connector = http://127.0.0.1:12345
debug
libdebug
# dinout
```

And post the extra debug messages that you will see in the output of `step-ca` and in the output of `yubihsm-connector`.

**Comment by mikansontap123** [comment]
Hi @maraino ,

Thanks for assisting.
I am also running the yubihsm2-sdk 2021.08 packages. 

I have also enabled the options in YUBIHSM_PKCS11_CONF as you mentioned. Perhaps notably the only difference is the HSM is listening remotely on another VM. I am able to run yubihsm-shell and interact with it from the same local machine that step-ca is running on, so I hope this shouldn't be related to the problem. 

I've made 3 attempts. I'll summarize them first before providing output.

**Attempt 1**: Requested step-ca to sign an SSH certificate, using OICD provisioner against Azure AD for authentication. Attempt is successful and a signed, valid certificate is returned.
`step ssh certificate certSubject id_ecdsa --ca-url https://my-step-ca.local --root root_ca.crt`

**Attempt 2**: Same command as above. Interestingly there is zero output from the yubihsm-connector service, not a single line. There is however a failure output coming from step-ca. 

**Attempt 3**: Same command as above. Now this time there is output from yubihsm-connector, with some errors. step-ca returns the same errors as Attempt 2. 

Outputs below. 

**Attempt 1**
yubihsm-connector output
```[P11 - INF 15:49:25.869426] yubihsm_pkcs11.c:844 (C_OpenSession): In
[P11 - INF 15:49:25.869487] yubihsm_pkcs11.c:889 (C_OpenSession): Allocated session 6
[P11 - INF 15:49:25.869490] yubihsm_pkcs11.c:891 (C_OpenSession): Out
[P11 - INF 15:49:25.869494] yubihsm_pkcs11.c:3469 (C_SignInit): In
[P11 - INF 15:49:25.869496] yubihsm_pkcs11.c:3494 (C_SignInit): Trying to sign data with mechanism 0x1041 and key 05039091
[P11 - INF 15:49:25.869502] yubihsm_pkcs11.c:3587 (C_SignInit): ECDSA signature requested
[P11 - INF 15:49:25.869504] yubihsm_pkcs11.c:3629 (C_SignInit): Out
[P11 - INF 15:49:25.869506] yubihsm_pkcs11.c:3642 (C_Sign): In
[P11 - INF 15:49:25.869507] yubihsm_pkcs11.c:3667 (C_Sign): The size of the signature will be 64
[P11 - INF 15:49:25.869508] yubihsm_pkcs11.c:3676 (C_Sign): Out
[P11 - INF 15:49:25.869509] yubihsm_pkcs11.c:3642 (C_Sign): In
[P11 - INF 15:49:25.869510] yubihsm_pkcs11.c:3667 (C_Sign): The size of the signature will be 64
[P11 - INF 15:49:25.869511] yubihsm_pkcs11.c:3690 (C_Sign): Sending 32 bytes to sign
[P11 - INF 15:49:25.869512] yubihsm_pkcs11.c:3704 (C_Sign): Using key 9091
[P11 - INF 15:49:25.869513] yubihsm_pkcs11.c:3705 (C_Sign): After padding and transformation there are 32 bytes
[LIB - NET 15:49:25.869516] yubihsm.c:440 (send_encrypted_msg):  SEND >> (03 + 0034) <redacted>
[LIB - NET 15:49:25.869533] yubihsm.c:352 (send_authenticated_msg):  SEND >> (03 + 0049) <redacted>
[LIB - NET 15:49:25.869541] yubihsm.c:289 (send_msg):  SEND >> (03 + 0057) <redacted>
[LIB - NET 15:49:25.950089] yubihsm.c:296 (send_msg):  RECV << (03 + 0089) <redacted>
[LIB - NET 15:49:25.950108] yubihsm.c:404 (send_authenticated_msg):  RECV << (03 + 0081) <redacted>
```
step-ca output
```
INFO[0007] duration="101.729µs" duration-ns=101729 fields.time="2022-05-19T15:49:23+09:00" method=GET name=ca path="/provisioners?limit=100" protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2ugsvljdbdqv24kv5g size=692 status=200 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
INFO[0009] duration="98.087µs" duration-ns=98087 fields.time="2022-05-19T15:49:25+09:00" method=GET name=ca path=/version protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2ugtfljdbdqv24kv60 size=34 status=200 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
INFO[0009] duration=85.668818ms duration-ns=85668818 fields.time="2022-05-19T15:49:25+09:00" method=POST name=ca ott=<removed token to cut down on length> path=/ssh/sign protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2ugtfljdbdqv24kv6g size=875 status=201 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
```
At this point, I have a signed and valid certificate.

**Attempt 2**
_No yubihsm-connector output_

step-ca output
```
INFO[0090] duration="75.053µs" duration-ns=75053 fields.time="2022-05-19T15:50:46+09:00" method=GET name=ca path="/provisioners?limit=100" protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2uhhnljdbdqv24kv70 size=692 status=200 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
INFO[0091] duration="18.291µs" duration-ns=18291 fields.time="2022-05-19T15:50:48+09:00" method=GET name=ca path=/version protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2uhi7ljdbdqv24kv7g size=34 status=200 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
ERRO[0091] duration=5.344778ms duration-ns=5344778 error="authority.SignSSH: error signing certificate: error signing certificate: pkcs11: 0x6: CKR_FUNCTION_FAILED" fields.time="2022-05-19T15:50:48+09:00" method=POST name=ca ott=<removed token to cut down on length> path=/ssh/sign protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2uhi7ljdbdqv24kv80 size=148 status=500 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
```

**Attempt 3**
yubihsm-connector output
```
[P11 - INF 15:50:48.296106] yubihsm_pkcs11.c:844 (C_OpenSession): In
[P11 - INF 15:50:48.296142] yubihsm_pkcs11.c:889 (C_OpenSession): Allocated session 7
[P11 - INF 15:50:48.296144] yubihsm_pkcs11.c:891 (C_OpenSession): Out
[P11 - INF 15:50:48.296148] yubihsm_pkcs11.c:3469 (C_SignInit): In
[P11 - INF 15:50:48.296150] yubihsm_pkcs11.c:3494 (C_SignInit): Trying to sign data with mechanism 0x1041 and key 05039091
[P11 - INF 15:50:48.296156] yubihsm_pkcs11.c:3587 (C_SignInit): ECDSA signature requested
[P11 - INF 15:50:48.296157] yubihsm_pkcs11.c:3629 (C_SignInit): Out
[P11 - INF 15:50:48.296160] yubihsm_pkcs11.c:3642 (C_Sign): In
[P11 - INF 15:50:48.296161] yubihsm_pkcs11.c:3667 (C_Sign): The size of the signature will be 64
[P11 - INF 15:50:48.296162] yubihsm_pkcs11.c:3676 (C_Sign): Out
[P11 - INF 15:50:48.296163] yubihsm_pkcs11.c:3642 (C_Sign): In
[P11 - INF 15:50:48.296164] yubihsm_pkcs11.c:3667 (C_Sign): The size of the signature will be 64
[P11 - INF 15:50:48.296165] yubihsm_pkcs11.c:3690 (C_Sign): Sending 32 bytes to sign
[P11 - INF 15:50:48.296166] yubihsm_pkcs11.c:3704 (C_Sign): Using key 9091
[P11 - INF 15:50:48.296167] yubihsm_pkcs11.c:3705 (C_Sign): After padding and transformation there are 32 bytes
[LIB - NET 15:50:48.296170] yubihsm.c:440 (send_encrypted_msg):  SEND >> (03 + 0034) <redacted>
[LIB - NET 15:50:48.296195] yubihsm.c:352 (send_authenticated_msg):  SEND >> (03 + 0049) <redacted>
[LIB - NET 15:50:48.296203] yubihsm.c:289 (send_msg):  SEND >> (03 + 0057) <redacted>
[LIB - NET 15:50:48.298177] yubihsm.c:296 (send_msg):  RECV << (03 + 0001) <redacted>
[LIB - NET 15:50:48.298183] yubihsm.c:404 (send_authenticated_msg):  RECV << (03 + 0001) <redacted>
[LIB - ERR 15:50:48.298184] yubihsm.c:474 (send_encrypted_msg): send_authenticated_msg Unable to verify MAC
[LIB - ERR 15:50:48.298187] yubihsm.c:1823 (yh_util_sign_ecdsa): Failed to send SIGN ECDSA command: Unable to verify MAC
[P11 - ERR 15:50:48.298188] yubihsm_pkcs11.c:3711 (C_Sign): Unable to sign data
[P11 - INF 15:52:03.095585] yubihsm_pkcs11.c:844 (C_OpenSession): In
[P11 - INF 15:52:03.095647] yubihsm_pkcs11.c:889 (C_OpenSession): Allocated session 8
[P11 - INF 15:52:03.095652] yubihsm_pkcs11.c:891 (C_OpenSession): Out
[P11 - INF 15:52:03.095656] yubihsm_pkcs11.c:3469 (C_SignInit): In
[P11 - INF 15:52:03.095658] yubihsm_pkcs11.c:3494 (C_SignInit): Trying to sign data with mechanism 0x1041 and key 05039091
[P11 - INF 15:52:03.095665] yubihsm_pkcs11.c:3587 (C_SignInit): ECDSA signature requested
[P11 - INF 15:52:03.095666] yubihsm_pkcs11.c:3629 (C_SignInit): Out
[P11 - INF 15:52:03.095668] yubihsm_pkcs11.c:3642 (C_Sign): In
[P11 - INF 15:52:03.095669] yubihsm_pkcs11.c:3667 (C_Sign): The size of the signature will be 64
[P11 - INF 15:52:03.095671] yubihsm_pkcs11.c:3676 (C_Sign): Out
[P11 - INF 15:52:03.095673] yubihsm_pkcs11.c:3642 (C_Sign): In
[P11 - INF 15:52:03.095685] yubihsm_pkcs11.c:3667 (C_Sign): The size of the signature will be 64
[P11 - INF 15:52:03.095686] yubihsm_pkcs11.c:3690 (C_Sign): Sending 32 bytes to sign
```

step-ca output
```
INFO[0164] duration="94.81µs" duration-ns=94810 fields.time="2022-05-19T15:52:01+09:00" method=GET name=ca path="/provisioners?limit=100" protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2ui4fljdbdqv24kv8g size=692 status=200 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
INFO[0166] duration="18.96µs" duration-ns=18960 fields.time="2022-05-19T15:52:03+09:00" method=GET name=ca path=/version protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2ui4vljdbdqv24kv90 size=34 status=200 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=
ERRO[0166] duration=5.80877ms duration-ns=5808770 error="authority.SignSSH: error signing certificate: error signing certificate: pkcs11: 0x6: CKR_FUNCTION_FAILED" fields.time="2022-05-19T15:52:03+09:00" method=POST name=ca ott=<removed token to cut down on length>path=/ssh/sign protocol=HTTP/2.0 referer= remote-address=10.12.0.155 request-id=ca2ui4vljdbdqv24kv9g size=148 status=500 user-agent="Smallstep CLI/0.19.0 (linux/amd64)" user-id=

```

I also repeated the above scenario with another 3 attempts, and interestingly there is also zero yubihsm-connector output on the attempt 2 as before. 

Given that the yubihsm-connector outputs for attempt 1 and attempt 2 start with `Allocated session 6` and `Allocated session 7`, it appears that there is no interaction going on between step-ca and yubihsm on attempt 3.

**Comment by maraino** [comment]
I'm wondering if you're using different versions for the `yubihsm2-sdk` and if that can cause any issue. I'm saying this because your logs do not match `2021.08`, they actually match `2021.12`, yubihsm-shell 2.3.0 instead of 2.2.0. 

See https://github.com/Yubico/yubihsm-shell/blob/2.3.0/lib/yubihsm.c

I will need to test this scenario with two different computers to see if it can be the root of any problem, can you make sure to run the same version on both servers to make sure this is not the cause.

**Comment by maraino** [comment]
@mikansontap123 I've just updated to 2.3.0 and I'm able to reproduce your issue. So right now seems to work fine with yubihsm-shell 2.2.0 (2021.08) but not 2.3.0 (2021.12)

**Comment by mikansontap123** [comment]
Thanks @maraino, sorry for the misleading info - looking back now, indeed we did upgrade yubihsm-shell due to reported security issues from Yubico [YSA-2021-04](https://www.yubico.com/support/security-advisories/ysa-2021-04/) . Coincidentally this advisory directly impacts SSH certificate signing. 

In the current setup, yubihsm-shell is 2.3.0 on the local VM (running step-ca) while yubihsm-shell is 2.2.0 on the remote VM (running the listener.) I will experiment with running both 2.2.0 and both 2.3.0 to observe the behavior on my end.

**Comment by maraino** [comment]
I've created an issue in yubihsm-shell to see if it's the expected behavior or a bug in their code. If it's the expected behavior I'll add an option to send some kind of keepalive, I've experimented with that and seems to work.

**Comment by maraino** [comment]
Hi @mikansontap123, it seems that it was a regression issue in yubihsm-shell that is fixed if you compile from the source code. See https://github.com/Yubico/yubihsm-shell/issues/262

I'll close this issue if they close the one in yubihsm-shell.

**Comment by mikansontap123** [comment]
@maraino 

Thanks a lot for reaching out to Yubico. I've installed yubihsm-shell 2.3.1 and can confirm the problem appears to be resolved now. I am now able to continuously issue certificates after the default keepalive has passed. 👍

As a side note, I've installed this only on the local machine running step-ca, and not the remote VM with the HSM attached, which is still running 2.2.0. This disparity doesn't seem to be causing any issue at the moment.

**Comment by maraino** [comment]
This issue is caused by a regression problem in yubihsm-shell 2.3.0, which is fixed in 2.3.1. Although there are no official releases for the SDK with the fix, there are official releases of yubihsm-shell with it:
https://developers.yubico.com/yubihsm-shell/Releases/
